core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/StringCodec.java

```diff
@@ -20,22 +20,36 @@
 import com.datastax.oss.driver.api.core.type.codec.TypeCodec;
 import com.datastax.oss.driver.api.core.type.reflect.GenericType;
 import com.datastax.oss.driver.internal.core.util.Strings;
-import com.datastax.oss.protocol.internal.util.Bytes;
 import edu.umd.cs.findbugs.annotations.NonNull;
 import edu.umd.cs.findbugs.annotations.Nullable;
 import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.charset.CharacterCodingException;
 import java.nio.charset.Charset;
+import java.nio.charset.CharsetDecoder;
+import java.nio.charset.CharsetEncoder;
+import java.nio.charset.CodingErrorAction;
 import net.jcip.annotations.ThreadSafe;
 
 @ThreadSafe
 public class StringCodec implements TypeCodec<String> {
 
   private final DataType cqlType;
-  private final Charset charset;
+  private final CharsetEncoder encoder;
+  private final CharsetDecoder decoder;
 
   public StringCodec(@NonNull DataType cqlType, @NonNull Charset charset) {
     this.cqlType = cqlType;
-    this.charset = charset;
+    this.encoder =
+        charset
+            .newEncoder()
+            .onMalformedInput(CodingErrorAction.REPORT)
+            .onUnmappableCharacter(CodingErrorAction.REPORT);
```

Comment:
Hm, starting to think this is wrong, because these encoders aren't thread-safe. Will update the PR in a bit

Reply:
It was definitely wrong, so I went with FastThreadLocal like in the TimestampCodec.